File: 130_Surrounded_Regions/solve.py
"""
Any non surrounded region must connect to a 'O' at the boundary.
Traverse all 'O' on the bounding and appply BFS on that.

"""
import collections
class Solution(object):
    def solve(self, board):
        """
        :type board: List[List[str]]
        :rtype: void Do not return anything, modify board in-place instead.
        """
        displacements = [[1, 0], [-1, 1], [-1, -1], [1, -1]]

        def bfs(i, j, m, n):
            dq = collections.deque([[i, j]])
            board[i][j] = 'B'
            while dq:
                i, j = dq.popleft()
                for di, dj in displacements:
                    i += di
                    j += dj
                    if 0 <= i < m and 0 <= j < n and board[i][j] == 'O':
                        dq.append([i, j])
                        board[i][j] = 'B'

        m, n = len(board), len(board[0])

        for i in xrange(m):
            for j in [0, n - 1]:
                if board[i][j] == 'O':
                    bfs(i, j, m, n)

        for i in [0, m - 1]:
            for j in xrange(n):
                if board[i][j] == 'O':
                    bfs(i, j, m, n)

        for i in xrange(m):
            for j in xrange(n):
                if board[i][j] != 'X':
                    board[i][j] = 'X' if board[i][j] == 'O' else 'O'

sol = Solution()
board = [list("XXXX"), list("XOOX"), list("XOOX"), list("XOXX")]
sol.solve(board)

File: 130_Surrounded_Regions/solve.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
    public:
        void solve(vector<vector<char> >& board) {
            int m = board.size();
            int n = board[0].size();

            for(int j = 0; j < n; j++) {
                if(board[0][j] == 'O') {
                    bfsBoundary(board, 0, j);
                }

                if(board[m - 1][j] == 'O') {
                    bfsBoundary(board, m - 1, j);

                }
            }

            for(int i = 0; i < m; i++) {
                if(board[i][0] == 'O' ) {
                    bfsBoundary(board, i, 0);
                }

                if(board[i][n - 1] == 'O') {
                    bfsBoundary(board, i, n - 1);
                }
            }

            for(int i = 0; i < m; i++) {
                for(int j = 0; j < n; j++) {
                    if(board[i][j] == 'B') {
                        board[i][j] = 'O';
                    }
                    else if(board[i][j] == 'O') {
                        board[i][j] = 'X';
                    }
                }
            }
        }
    private:
        void bfsBoundary(vector<vector<char> >& board, int w, int l) {
            int m = board.size();
            int n = board[0].size();
            int dw[4] = {1, -1, 0, 0};
            int dl[4] = {0, 0, 1, -1};
                
            board[w][l] = 'B';

            for(int i = 0; i < 4; i++) {
                int nextW = w + dw[i];
                int nextL = l + dl[i];
                if(nextW >= 0 && nextW < m &&
                   nextL >= 0 && nextL < n &&
                   board[nextW][nextL] == 'O') {
                    bfsBoundary(board, nextW, nextL);
                }
            }

        }
};
